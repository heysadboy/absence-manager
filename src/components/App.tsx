import { useState, useEffect } from 'react';
import absenceData from '../data/absences.json';
import memberData from '../data/members.json';
import { IMemberList, IAbsence } from '../util/interfaces';
import { TPageIndex } from '../util/types';
import AbsenceList from './AbsenceList';
import PaginationFooter from './PaginationFooter';
import _ from 'lodash';

enum EStatus {
    loading = 'loading',
    error = 'error',
    success = 'success'
}

const App = () => {
    const [members, setMembers] = useState<IMemberList>({});
    const [absences, setAbsences] = useState<IAbsence[]>([]);
    const [status, setStatus] = useState<EStatus>(EStatus.loading);
    const [pageIndex, setPageIndex] = useState<TPageIndex>({ start: 0, end: 0, total: 0 });

    useEffect(() => {
        const pageTotal: number = absenceData.payload.length;
        if (pageTotal < 10) {
            setPageIndex({ start: 0, end: pageTotal, total: pageTotal });
        }
        else {
            setPageIndex({ start: 0, end: 10, total: pageTotal })
        }
    }, []);

    useEffect(() => {
        if (memberData.message === "Success" && absenceData.message === "Success") {
            setMembers({ ..._.mapKeys(memberData.payload, 'userId') });
            setAbsences(absenceData.payload.slice(pageIndex.start, pageIndex.end));
            setStatus(EStatus.success);
        }
        else {
            setStatus(EStatus.error);
        }
    }, [status, pageIndex]);

    if (status === EStatus.loading) {
        return (<div>Loading</div>);
    }
    else if (status === EStatus.error) {
        return (<div>Error</div>);
    }
    else {
        return (
            <div className="ui container">
                <h1>Absence Manager</h1>
                <AbsenceList members={members} absences={absences} />
                <PaginationFooter pageIndex={pageIndex} setPageIndex={setPageIndex} />
            </div>
        );
    }
};

export default App;