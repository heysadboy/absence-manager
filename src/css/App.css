#error-message, #empty-message {
    margin-top: 150px;
}