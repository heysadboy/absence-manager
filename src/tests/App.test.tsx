import renderer from 'react-test-renderer';
import App from '../components/App';

test('App.tsx snapshot testing', () => {
    const component = renderer.create(
        <App />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});