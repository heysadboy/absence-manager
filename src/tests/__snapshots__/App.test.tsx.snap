// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.tsx snapshot testing 1`] = `
<div
  className="ui container"
  id="app-container"
>
  <div
    className="ui top fixed vertical menu"
    id="header-bar"
  >
    <p
      id="header-title"
    >
      Absence Manager
    </p>
    <div
      className="ui action input "
      id="date-filter"
    >
      <input
        className="ui input"
        onChange={[Function]}
        placeholder="yyyy-mm-dd - yyyy-mm-dd"
        value=""
      />
      <button
        className="ui  icon button"
        onClick={[Function]}
      >
        <i
          className="calendar icon"
        />
      </button>
    </div>
    <div
      className="ui simple selection dropdown"
      id="type-filter"
    >
      <i
        className="dropdown icon"
      />
      <div
        className="text"
      >
        All absence
      </div>
      <div
        className="menu"
      >
        <div
          className="active item"
          data-value="all"
          onClick={[Function]}
        >
          All absence
        </div>
        <div
          className=" item"
          data-value="sickness"
          onClick={[Function]}
        >
          Sickness
        </div>
        <div
          className=" item"
          data-value="vacation"
          onClick={[Function]}
        >
          Vacation
        </div>
      </div>
    </div>
  </div>
  <div
    className="ui bottom fixed menu"
    id="footer"
  >
    <div
      className="ui container"
    >
      <div
        id="footer-text"
      >
        <p>
          Showing 
          <b>
            1
          </b>
           - 
          <b>
            0
          </b>
           of 
          <b>
            0
          </b>
           absentees
        </p>
      </div>
      <div
        className="right menu"
      >
        <div
          className="ui basic buttons"
        >
          <button
            className="ui button"
            onClick={[Function]}
          >
            <i
              className="left chevron icon"
            />
          </button>
          <button
            className="ui button"
            onClick={[Function]}
          >
            <i
              className="right chevron icon"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    className="ui active large loader"
  />
</div>
`;
